t/plugin/referer-restriction.t

```diff
@@ -0,0 +1,189 @@
+                                     "*.xx.com",
+                                     "yy.com"
+                                 ]
+                            }
+                        }
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+
+
+
+=== TEST 2: hit route and in the whitelist (wildcard)
+--- request
+GET /hello
+--- more_headers
+Referer: http://www.xx.com
+--- response_body
+hello world
+
+
+
+=== TEST 3: hit route and in the whitelist
+--- request
+GET /hello
+--- more_headers
+Referer: https://yy.com/am
+--- response_body
+hello world
+
+
+
+=== TEST 4: hit route and not in the whitelist
+--- request
+GET /hello
+--- more_headers
+Referer: https://www.yy.com/am
+--- error_code: 403
+
+
+
+=== TEST 5: hit route and without Referer
+--- request
+GET /hello
+--- error_code: 403
+
+
+
+=== TEST 6: set whitelist, allow Referer missing
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "uri": "/hello",
+                        "upstream": {
+                            "type": "roundrobin",
+                            "nodes": {
+                                "127.0.0.1:1980": 1
+                            }
+                        },
+                        "plugins": {
+                            "referer-restriction": {
+                                "optional": true,
+                                 "whitelist": [
+                                     "*.xx.com",
+                                     "yy.com"
+                                 ]
+                            }
+                        }
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+
+
+
+=== TEST 7: hit route and without Referer
+--- request
+GET /hello
+--- response_body
+hello world
+
+
+
+=== TEST 8: malformed Referer is treated as missing
+--- request
+GET /hello
+--- more_headers
+Referer: www.yy.com
+--- response_body
+hello world
+
+
+
+=== TEST 9: invalid schema
+--- config
+    location /t {
+        content_by_lua_block {
+            local plugin = require("apisix.plugins.referer-restriction")
+            local cases = {
+                "x.*",
+                "~y.xn",
+                "::1",
+            }
+            for _, c in ipairs(cases) do
+                local ok, err = plugin.check_schema({
+                    whitelist = {c}
+                })
+                if ok then
+                    ngx.log(ngx.ERR, c)
+                end
+            end
+        }
+    }
+--- request
+GET /t
```

Comment:
has no assert here？

Reply:
@nic-chen 
All tests under this file already check the error log.

Reply:
OK